### PR TITLE
Unsaved Changes

### LINKED ISSUE #479: [FE] Unsaved outcomes
From this comment https://github.com/gnosis/conditional-tokens-explorer/pull/470#pullrequestreview-509582824

```
Unsaved outcomes: (this is more of a question or a dilemma)

- Add some outcomes.
- Click on the "Edit button" in all of them, change the text. (Do not save any changes.
- Here's the catch: should we send the form anyways - unsaved changes and all- when clicking on "Prepare"? Should we warn the user that there are unsaved changes? Or do we disable the "Prepare" button until all changes are saved?

```

My preference is this option 

```
Should we warn the user that there are unsaved changes?
```

### PR DESCRIPTION
**Pull request by hpmaxi**
Closes #479 

A message of unsaved changes is shown below the outcomes block only when everything else is completed. I think  we can
- Change the message to live inside `AddOutcome` 
- Always show the message, and use the proper component according to their final position
- Maybe we can just disable the submit, or making something more subtle.

Also, I wasn't sure if the possibility of editing multiple outcomes at the same time is desirable or not.

### CODE REVIEW

File: src/components/form/AddOutcome/index.tsx [hunk 1]
```diff
@@ -189,6 +204,7 @@ const EditableOutcome: React.FC<{
           onClick={() => {
             removeOutcome()
             setIsEditing(false)
+            onEditOutcome(false, outcomeIndex)
```
Comment (hpmaxi) [inline src/components/form/AddOutcome/index.tsx:207]: This isn't DRY but using an effect over `isEditing` generate race conditions.

File: src/pages/PrepareCondition/index.tsx [hunk 2]
```diff
@@ -162,21 +162,38 @@ export const PrepareCondition = () => {
...
+
+  const toggleEditOutcome = useCallback(
+    (value: boolean, index: number) => {
+      setOutcomesBeingEdited([
+        ...outcomesBeingEdited.slice(0, index),
+        value,
+        ...outcomesBeingEdited.slice(index + 1),
+      ])
+    },
+    [outcomesBeingEdited]
```
Comment (hpmaxi) [inline src/pages/PrepareCondition/index.tsx:196]: Maybe, a better way to manage this is using an array of `Outcome = { value: string, editing: boolean}`

File: src/pages/PrepareCondition/index.tsx [hunk 3]
```diff
@@ -737,8 +758,14 @@ export const PrepareCondition = () => {
...
               outcomes={outcomes}
               removeOutcome={removeOutcome}
+              toggleEditOutcome={toggleEditOutcome}
               updateOutcome={updateOutcome}
             />
+            {areOutcomesBeingEdited && !submitDisabled && (
+              <ErrorContainer>
+                <ErrorMessage>Unsaved changes</ErrorMessage>
+              </ErrorContainer>
+            )}
```
Comment (hpmaxi) [inline src/pages/PrepareCondition/index.tsx:768]: This and everything inside `PrepareCondition`  could be moved inside if we only show a message independent of the submit button state.

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr811--conditionaltokensexplorer.review.gnosisdev.com

**Review by gabitoesmiapodo — CHANGES_REQUESTED** [review]
I think this could be better if you:

- Disable the "Prepare" button until every outcome is saved (that is: everything should be OK before we allow the user to send the form, as it is everywhere else).
- Only when / if the outcome being edited loses focus show a warning. Use the proper warning component here, as the one below "Resolution Date". Place it below the "Outcomes" component.
- Maybe allow only one outcome to be edited at once? There could a few problems if we do this, though: user edited outcome 1, clicks "Edit" on outcome 2... what should we do? Discard edited value on outcome 1? Save edited value on outcome 1? The other option is to hide every edit / remove button except the ones we are working with. I'm not clear on what's the best option here, but probably the last one is easier to implement.

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr811--conditionaltokensexplorer.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr811--conditionaltokensexplorer.review.gnosisdev.com

**Comment by gabitoesmiapodo** [comment]
It's better now.

I think we could omit this point "Only when / if the outcome being edited loses focus show a warning. Use the proper warning component here, as the one below "Resolution Date". Place it below the "Outcomes" component." 

Don't show any message, just leave the "Prepare" button disabled. 

If you feel adventurous, you could change the outcome's input border color to red when it loses focus and it's unsaved (not before it loses focus, it should stay green until then).

**Review by gabitoesmiapodo — CHANGES_REQUESTED** [review]
See https://github.com/gnosis/conditional-tokens-explorer/pull/811#issuecomment-767799596

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr811--conditionaltokensexplorer.review.gnosisdev.com

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr811--conditionaltokensexplorer.review.gnosisdev.com

**Comment by hpmaxi** [comment]
@gabitoesmiapodo Thanks for all the feedback! It was very helpful. 

> If you feel adventurous, you could change the outcome's input border color to red when it loses focus and it's unsaved (not before it loses focus, it should stay green until then).

I did this, there is a second when it shows a red border color just before passing changing isEditing to false but losing focus (clicking on the check mark or the cross mark). I tried to move the `onBlur` callback to the `OutcomeWrapper` but I get the same result.

**Comment by elena-zh** [comment]
Hi @hpmaxi !
The changes look good to me, however, I have one question. How can user know that something is wrong with an Outcomes section when he leaves unsaved outcome at the bottom of the list (or at the top) and scrolls up/down? 
I see only 'Prepare' button disabled, however, everything is filled in in a proper way..
https://drive.google.com/file/d/1pna9fEHkztg1d23ikln57bZZHT4WD1eB/view
From my perspective, it would be nice to show a 'visible' warning for the outcomes section. At least, colour it in red. 
What do you think?

**Comment by hpmaxi** [comment]
@elena-zh I'm not sure, I think this section is short so we can be a little less explicit about things the user just hides (scrolling) a few seconds before, but that just my opinion, I'll wait for the reviewers and maybe if they approve this we can improve it later.

**Review by gabitoesmiapodo — APPROVED** [review]
> I did this, there is a second when it shows a red border color just before passing changing isEditing to false but losing focus (clicking on the check mark or the cross mark). I tried to move the onBlur callback to the OutcomeWrapper but I get the same result.

It looks fine, we can improve that little thing later.

In this case, an unsaved outcome is the only non obvious form error, so the red line / number helps the user to find the problem. No need for extra messages.

**Pull request merged**